apps/browser/src/content/autofill.js

```diff
@@ -31,19 +31,110 @@
-  9. Add new handler, for new command that responds with page details in response callback
+  1.  Populate isFirefox
+  2.  Remove isChrome and isSafari since they are not used.
+  3.  Unminify and format to meet Mozilla review requirements.
+  4.  Remove unnecessary input types from getFormElements query selector and limit number of elements returned.
+  5.  Remove fakeTested prop.
+  6.  Rename com.agilebits.* stuff to com.bitwarden.*
+  7.  Remove "some useful globals" on window
+  8.  Add ability to autofill span[data-bwautofill] elements
+  9.  Add new handler, for new command that responds with page details in response callback
   10. Handle sandbox iframe and sandbox rule in CSP
   11. Work on array of saved urls instead of just one to determine if we should autofill non-https sites
+  12. Implement new HTML element query logic to be able to traverse into ShadowRoot
   */
 
+  /*
+   * `openOrClosedShadowRoot` is only available to WebExtensions.
+   * We need to use the correct implementation based on browser.
+   */
+  // START MODIFICATION
+  var getShadowRoot;
+
+  if (chrome.dom && chrome.dom.openOrClosedShadowRoot) {
+      // Chromium 88+
+      // https://developer.chrome.com/docs/extensions/reference/dom/
+      getShadowRoot = function (element) {
+          if (!(element instanceof HTMLElement)) {
+              return null;
+          }
+
+          return chrome.dom.openOrClosedShadowRoot(element);
+      };
+  } else {
+      getShadowRoot = function (element) {
+          // `openOrClosedShadowRoot` is currently available for Firefox 63+
+          // https://developer.mozilla.org/en-US/docs/Web/API/Element/openOrClosedShadowRoot
+          // Fallback to usual shadowRoot if it doesn't exist, which will only find open ShadowRoots, not closed ones.
+          // https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot#browser_compatibility
+          return element.openOrClosedShadowRoot || element.shadowRoot;
+      };
+  }
+
+  /*
+   * Works like querySelectorAll but uses a filerCallback where the node is passed to instead of a CSS selector string.
+   * It has the advantage over querySelectorAll that it traverses into ShadowRoot.
+   */
+  function queryDocAll(doc, rootEl, filterCallback, els) {
+      els = els || [];
+
+      if(typeof filterCallback !== 'function') {
+          filterCallback = function () { /* noop */ };
+      }
+
+      var treeWalker = doc.createTreeWalker(rootEl, NodeFilter.SHOW_ELEMENT);
+      var node;
+
+      while (node = treeWalker.nextNode()) {
+        if (filterCallback(node)) {
+            els.push(node);
+        }
+
+        // If node contains a ShadowRoot we want to step into it and also traverse all child nodes inside.
+        var nodeShadowRoot = getShadowRoot(node);
+        if (nodeShadowRoot) {
+            // recursively traverse into ShadowRoot
+            els = queryDocAll(doc, nodeShadowRoot, filterCallback, els);
```

Comment:
Doesn't matter much, but I am not a fan of reassigning the value that we also mutate in the called recursion. Can we not just have queryDocAll return its own array, and when returning from recursing, extend ours with the results?

```suggestion
            els.push(...queryDocAll(doc, nodeShadowRoot, filterCallback));
```

Reply:
I also prefer to avoid mutation here. I think that `concat`ing the arrays might be faster than the spread operator (given that it's purpose built for combining 2 arrays, which is all we're doing here) - but I don't have any data to back that up. Either is probably fine.

I take your point about performance, but I'm not too concerned about it here. We can benchmark later if it's an issue.

Reply:
I built a little benchmark over here: https://jsbench.me/fhlb3cqqj2/1

My results in Firefox 107 on Linux Mint `Mozilla/5.0 (X11; Linux x86_64; rv:107.0) Gecko/20100101 Firefox/107.0` are:
```
Array.concat
3846427.63 ops/s ± 0.53%
90.71 % slower

Array.push with spread operator
10707846.38 ops/s ± 0.23%
74.14 % slower

Array.push directly onto elements
41413583.66 ops/s ± 1.18%
Fastest
```

So for a DOM with a lot of elements and many ShadowRoots, as can happen when using https://ionicframework.com/ as an example, the direct `Array.push` has a significant performance- and probably also memory benefit.

A reason why I really care about performance is that I have to use the Dashlane Password Manager for work and on pages with a lot of inputs it often happens that the page gets very slow. After some seconds Firefox shows a toolbar at the top of the page saying something like "Dashlane is slowing down your browser, would you like to stop it?"

Maybe we can add a comment explaining why I implemented it like that instead?

Reply:
Resolved by af893d3 / 3dc542b
Thanks @chadm-sq!

Reply:
Good point! Kudos for taking the time to benchmark.

Reply:
BTW this is the message I was talking about. Happened on a page with 1110 input fields (Shopware 6 detailed role privileges page).

![image](https://user-images.githubusercontent.com/14234815/205124557-36380222-7dd6-4754-bca1-403183526945.png)

Reply:
I'll be keen to know if bitwarden gives a similar warning, @RafaelKr .

Ooo, optimization for the future? We don't need to scan the entire document. I bet the displayed and on-screen (p1 < .top < p2) elements only matter. If we can't see them, we don't need to try to fill them.

Reply:
Nope, never seen this one for Bitwarden. Personally I used LastPass before Bitwarden, also I've never seen it for that.
I'm not sure what Dashlane is doing to produce this warning, I don't know how this would perform on a low-end device.

I also don't like Dashlane at all, many Features are implemented in a bad way or not at all. Maybe we'll switch to Bitwarden for work when we have some time to do so ;)

So you mean we should also check for hidden elements and don't traverse them? Maybe a good idea, but only after implementing the mutation observer. Because many pages can toggle between login and register forms and maybe for Single Page Applications Bitwarden wouldn't recognize that change.